### ISSUE TITLE
About the training process

### ISSUE THREAD

**Issue by skr998** [frank-xwang/CLD-UnsupervisedLearning#22]
Hi, thanks for your excellent work!

There are some problems about the training process after reading the code:
1.What role does DistributedShufle.forward_shuffle() and DistributedShufle.backward_shuffle() play here? Can't you directly set shuffle=True in the dataloader? Is it because of the use of distributed training mode?
![image](https://user-images.githubusercontent.com/77918488/167248208-d5ba2d95-7371-4fab-9ab0-9d3d58d16c97.png)

2.I found you built three augmented samples (x1, x2, x3) for x, why not use only two augmented samples like MoCo? Isn't the calculation of general contrastive loss in this form "[L(f(x1), f(x2))+L(f(x2), f(x1))]/2" in terms of the positive pair {x1, x2}?

3.The paper mentioned "memory bank v is computed as the average feature of all the augmented versions of x_i seen so far". 
![image](https://user-images.githubusercontent.com/77918488/167249024-b56b7f97-ed04-400b-996f-9bbb13cbea4f.png)
How to understand the meaning of "average feature" here? I don't seem to be able to find the relevant implementation in the code.

I am looking forward to your answers. Thank you very much!

**Comment by frank-xwang** [comment]
Hi, sorry for the late reply. 
1) The functionalities of forward_shuffle and backward_shuffle are similar to the _batch_shuffle_ddp and _batch_unshuffle_ddp functions in [MoCo v2](https://github.com/facebookresearch/moco/blob/main/moco/builder.py#L69). The purpose of the shuffle BN operation is to prevent information leakage from intra-batch communication between samples (caused by BN), which may cause model cheating. Please refer to [MoCo paper Section 3.3](https://arxiv.org/pdf/1911.05722.pdf) for more details.
2) The cross-view comparison is mainly for stabilizing the model optimization process and increase the model convergence speed. Please refer to [CLD section 3](https://arxiv.org/pdf/2008.03813.pdf) for more details.
3) If we are talking about MoCo v2 + CLD, the prototype in the memory is generated by the EMA model with exponential moving averaged weights.
I hope this helps! If you have further questions, please feel free to reopen this issue.
